### ISSUE TITLE
Proposal: Improved Browser Connectivity using Secure WebSockets

### ISSUE THREAD

**Issue by marten-seemann** [libp2p/go-libp2p#1331]
This is a proposal to make it easier to use secure websockets (wss). Currently, users wishing to offer wss to the network need to run an HTTPS server (e.g. an nginx) with a valid TLS certificate configured in front of their libp2p node.

## Benefits

Browser nodes have limited options to connect to the rest of the libp2p network. There's no way to establish raw TCP or QUIC connections, and WebSockets / WebTransport connections are subject to the Web Security Model (i.e., the server has to present a CA-signed TLS certificate).
The only exception to this rule is WebRTC, which uses self-signed certificates. However, browsers enforce pretty strict limits on the number of concurrent WebRTC connection.

## Limitations

It is understood that this is only an option for public nodes, and only if they have a domain name configured. Therefore, only some of the public nodes offer wss support will actually make use of wss. However, this is still a lot better than the status quo, as this is an _additional_ (and not as strictly limited as WebRTC) connection option for libp2p nodes.

## Obtaining a Certificate

There are 3 options to get a valid certificate:
1. Import an existing key / cert file. This is probably not used very commonly nowadays.
2. Use the ACE HTTP or TLS challenge. Note that this involves starting a HTTP or endpoint on port 80 or 443, respectively (this is a limitation imposed by the ACME protocol). Note that this will pose a problem for nodes that run a HTTP(S) server on their node, or nodes which can't bind to ports < 1024.
3. Use DNS challenges. This will require user configuration, as DNS entries have to be modified temporarily to pass the ACME challenge (see the [certmagic documentation](https://github.com/caddyserver/certmagic#dns-challenge) for details)

In practice, it's only possible to obtain certificates for domain names, but not for IP addresses. It thus only makes sense to dial `/dnsaddr/domain.xyz/wss` addresses. Dialing `/ipX/<ip>/tcp/<port>/wss` addresses can still make sense for non-browser use cases, where we can ignore certificate validation errors (or, if we want to be fancy, distinguish between these two uses cases and use a libp2p-tls certificate in that case).

## Implementation Details

`go-ws-transport` will add a new configuration option `WithCertificate(*tls.Config)`. When set, it is possible to listen on `/wss` addresses.

[certmagic](https://github.com/caddyserver/certmagic) looks like it has everything we need, and we can offer all of the 3 options listed above to configure the certificate.

The domain names that we obtain certificates for could be configured manually, or extracted them from `/dnsaddr/` multiaddrs. It is unclear if this logic should be part of libp2p, or left to the application.

cc @willscott @vyzo @aschmahmann @aarshkshah1992 @mxinden @Stebalien @patrickwoodhead

**Comment by achingbrain** [comment]
This is going to solve a lot of problems for browser nodes 👍 

Tangentially related but I put a POC together a while back that does a reverse-DNS lookup on your router's external interface and uses uPNP to configure port forwarding for long enough to do the ACME handshake based on that domain name - https://gist.github.com/achingbrain/8d9a00190f2e13182579d26d138f5b4a - It also means you don't need to run the server on privileged ports.

**Comment by BigLep** [comment]
There is also https://github.com/protocol/web3-dev-team/pull/70 from 2021.

**Comment by lidel** [comment]
There are also some notes in https://github.com/ipfs/in-web-browsers/issues/181, 
TLDR being the various degrees of automation:
- **Stage 0**:  bring your own certs, manual config (issuance and renewal happens outside)
- **Stage 1**:  bring your own domain, manual initial config (then issuance and renewal are automated via ACME challenges)
- **Stage 2**: TLS out-of-the-box, no manual config (either PL provides a subdomain, or we do reverse-DNS lookup trick, or both depending on network topoligy, but the end goal is that user does nothing)

`/wss` story has been blocked for years because we were aiming at full automation (2), which comes with some caveats. I'm :+1: for doing this one step at a time, to keep the scope in check.

**I suggest we focus on (0) first,** and allow users to provide certs for domains  via config:
- ACME convenient, and popular, but there will always be security-sensitive deployments which prefer to do own key and cert  management. 
- Small scope, easy to test end-to-end with self-signed certs.

When we have this, we could look into **(1)** and **(2)**.

(I'm :+1: for  having built-in ACME support  but that usually baloons complexity across stack. This is because when we have ACME, full automation **(2)** is _very_ close. For example,  [PoC  by @achingbrain](https://gist.github.com/achingbrain/8d9a00190f2e13182579d26d138f5b4a) is worth exploring, as it would work in many home NATs  – the publicly diallable IP could be read via AutoNAT, and thanks to liberal defaults in consumer routers  we could attempt to map public  port `80` to local port `>1024` for HTTP Challenge, and fall-back to a subdomain and DNS-challenge via centralized infra provided by PL)

**Comment by marten-seemann** [comment]
> either PL provides a subdomain, or we do reverse-DNS lookup trick, or both depending on network topoligy, but the end goal is that user does nothing

I agree this should be out of scope. Providing a subdomain brings us into all kinds of trouble, first and foremost ACME rate-limiting. Reverse DNS seems straightforward enough, but will potentially require some more wiring up into the libp2p stack (just obtaining the cert is only the first half, we also need to advertise a `/dns/<domain>/wss` address).

Assuming a wss listener support in go-ws-transport (I have a working prototype that shouldn't be too far from being shippable), both **(0)** and **(1)** seem pretty straightforward (this is mostly due to the fact that certmagic is such a terrific library).

I created a small wrapper in https://github.com/marten-seemann/go-libp2p-certbot/. It supports both options:
* **(0)**: it's 1 LOC, see https://github.com/marten-seemann/go-libp2p-certbot/blob/c6961b3edd687b2735bea526c3fb549f6dd673c7/certmanager.go#L123-L129
* **(1)**: given a list of (`/dns`) multiaddrs, the library performs ACME for every domain. I'd call this semi-automatic, since the user only has to configure its listen addrs.

The `tls.Config` obtained from go-libp2p-certbot can directly be plugged into go-ws-transport.

---
@achingbrain's POC looks very interesting, but will probably require us to assign (sub)domains, right? Home IP addresses change very frequently, so we'd probably need a way to quickly obtain a new domain name. That's a whole lot more complexity, and I'd consider this out-of-scope for my proposal here. Let's focus on public nodes with pre-configured domains for now.

**Comment by marten-seemann** [comment]
Closing, since we now have WebTransport and that's 100x better than WebSocket. Specifically, it doesn't require any DNS setup.
